{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/hpc/users/restrp01/.local/lib/python3.10/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.\n",
      "  warnings.warn(\n",
      "/hpc/users/restrp01/.local/lib/python3.10/site-packages/xarray_schema/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import DistributionNotFound, get_distribution\n",
      "/hpc/users/restrp01/.local/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import tangram as tg\n",
    "\n",
    "from scipy import sparse\n",
    "import gc\n",
    "import os\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.colors as clr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "torch.set_float32_matmul_precision(\"high\")\n",
    "torch.cuda.is_available()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import sparse\n",
    "from scipy import io\n",
    "import os\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "merfish = sc.read_h5ad(\"data/ns-atlas.merfish_baysor.scanvi_integrated.cellcharter.anndata.annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "merfish.X = merfish.layers['counts'].copy()\n",
    "sc.pp.normalize_total(merfish)\n",
    "merfish.layers['norm'] = merfish.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39681"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merfish.var_names = merfish.var_names.str.lower().values.tolist()\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For training, we want to make sure that we train with the same genes across all samples. We will run each imaging batch separately, so make sure the genes are detected not only across gene panels but have at least one count across all batches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "merfish_batches = np.unique(merfish.obs['batch'].values.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(merfish, layer='log1p', n_top_genes=len(merfish.var_names), batch_key='batch', subset=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_panel</th>\n",
       "      <th>pathways</th>\n",
       "      <th>cell_category</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cell_type.detailed</th>\n",
       "      <th>Notes</th>\n",
       "      <th>in_both_panels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aadac</th>\n",
       "      <td>0.001948</td>\n",
       "      <td>0.038527</td>\n",
       "      <td>True</td>\n",
       "      <td>0.003182</td>\n",
       "      <td>0.292109</td>\n",
       "      <td>-0.302599</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>NSv2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Fib</td>\n",
       "      <td>Fib</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abi3bp</th>\n",
       "      <td>0.042268</td>\n",
       "      <td>0.181868</td>\n",
       "      <td>True</td>\n",
       "      <td>0.068665</td>\n",
       "      <td>0.502537</td>\n",
       "      <td>-0.315946</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ABI3BP</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Epithelia</td>\n",
       "      <td>Ecc Gland</td>\n",
       "      <td>Ecc Gland</td>\n",
       "      <td>Ecc gland</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acan</th>\n",
       "      <td>0.001811</td>\n",
       "      <td>0.035629</td>\n",
       "      <td>True</td>\n",
       "      <td>0.002572</td>\n",
       "      <td>0.334071</td>\n",
       "      <td>-0.518449</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ACAN</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>DS</td>\n",
       "      <td>DS</td>\n",
       "      <td>DS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ackr1</th>\n",
       "      <td>0.059810</td>\n",
       "      <td>0.274727</td>\n",
       "      <td>True</td>\n",
       "      <td>0.124364</td>\n",
       "      <td>1.183953</td>\n",
       "      <td>1.171005</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ACKR1</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>EC</td>\n",
       "      <td>EC</td>\n",
       "      <td>EC</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ackr2</th>\n",
       "      <td>0.009997</td>\n",
       "      <td>0.087661</td>\n",
       "      <td>True</td>\n",
       "      <td>0.016648</td>\n",
       "      <td>0.498767</td>\n",
       "      <td>0.034042</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ACKR2</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>EC</td>\n",
       "      <td>EC</td>\n",
       "      <td>EC</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wnt5a</th>\n",
       "      <td>0.015567</td>\n",
       "      <td>0.101244</td>\n",
       "      <td>True</td>\n",
       "      <td>0.022572</td>\n",
       "      <td>0.371165</td>\n",
       "      <td>-0.389344</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>WNT5A</td>\n",
       "      <td>Both</td>\n",
       "      <td>Wnt L-R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wnt7b</th>\n",
       "      <td>0.008498</td>\n",
       "      <td>0.082435</td>\n",
       "      <td>True</td>\n",
       "      <td>0.013407</td>\n",
       "      <td>0.449020</td>\n",
       "      <td>-0.124062</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>WNT7B</td>\n",
       "      <td>Both</td>\n",
       "      <td>Wnt L-R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xcl1</th>\n",
       "      <td>0.001085</td>\n",
       "      <td>0.027053</td>\n",
       "      <td>True</td>\n",
       "      <td>0.001564</td>\n",
       "      <td>0.298894</td>\n",
       "      <td>-0.644429</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>XCL1</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Immune</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>znf331</th>\n",
       "      <td>0.039304</td>\n",
       "      <td>0.168637</td>\n",
       "      <td>True</td>\n",
       "      <td>0.057845</td>\n",
       "      <td>0.491066</td>\n",
       "      <td>-0.215072</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ZNF331</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Immune</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>T cell (RGCC)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>znf442</th>\n",
       "      <td>0.004526</td>\n",
       "      <td>0.054851</td>\n",
       "      <td>True</td>\n",
       "      <td>0.006471</td>\n",
       "      <td>0.293572</td>\n",
       "      <td>-0.656928</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>ZNF442</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stroma</td>\n",
       "      <td>Peri</td>\n",
       "      <td>Peri</td>\n",
       "      <td>Pericyte III</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>562 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            mean       std  highly_variable     means  dispersions  \\\n",
       "aadac   0.001948  0.038527             True  0.003182     0.292109   \n",
       "abi3bp  0.042268  0.181868             True  0.068665     0.502537   \n",
       "acan    0.001811  0.035629             True  0.002572     0.334071   \n",
       "ackr1   0.059810  0.274727             True  0.124364     1.183953   \n",
       "ackr2   0.009997  0.087661             True  0.016648     0.498767   \n",
       "...          ...       ...              ...       ...          ...   \n",
       "wnt5a   0.015567  0.101244             True  0.022572     0.371165   \n",
       "wnt7b   0.008498  0.082435             True  0.013407     0.449020   \n",
       "xcl1    0.001085  0.027053             True  0.001564     0.298894   \n",
       "znf331  0.039304  0.168637             True  0.057845     0.491066   \n",
       "znf442  0.004526  0.054851             True  0.006471     0.293572   \n",
       "\n",
       "        dispersions_norm  highly_variable_nbatches  \\\n",
       "aadac          -0.302599                        17   \n",
       "abi3bp         -0.315946                        25   \n",
       "acan           -0.518449                        25   \n",
       "ackr1           1.171005                        25   \n",
       "ackr2           0.034042                        25   \n",
       "...                  ...                       ...   \n",
       "wnt5a          -0.389344                        25   \n",
       "wnt7b          -0.124062                        25   \n",
       "xcl1           -0.644429                        25   \n",
       "znf331         -0.215072                        25   \n",
       "znf442         -0.656928                        25   \n",
       "\n",
       "        highly_variable_intersection gene_symbol gene_panel pathways  \\\n",
       "aadac                          False       AADAC       NSv2      NaN   \n",
       "abi3bp                          True      ABI3BP       Both      NaN   \n",
       "acan                            True        ACAN       Both      NaN   \n",
       "ackr1                           True       ACKR1       Both      NaN   \n",
       "ackr2                           True       ACKR2       Both      NaN   \n",
       "...                              ...         ...        ...      ...   \n",
       "wnt5a                           True       WNT5A       Both  Wnt L-R   \n",
       "wnt7b                           True       WNT7B       Both  Wnt L-R   \n",
       "xcl1                            True        XCL1       Both      NaN   \n",
       "znf331                          True      ZNF331       Both      NaN   \n",
       "znf442                          True      ZNF442       Both      NaN   \n",
       "\n",
       "       cell_category  cell_type cell_type.detailed          Notes  \\\n",
       "aadac         Stroma        Fib                Fib            NaN   \n",
       "abi3bp     Epithelia  Ecc Gland          Ecc Gland      Ecc gland   \n",
       "acan          Stroma         DS                 DS             DS   \n",
       "ackr1         Stroma         EC                 EC             EC   \n",
       "ackr2         Stroma         EC                 EC             EC   \n",
       "...              ...        ...                ...            ...   \n",
       "wnt5a            NaN        NaN                NaN            NaN   \n",
       "wnt7b            NaN        NaN                NaN            NaN   \n",
       "xcl1          Immune         NK                 NK            NaN   \n",
       "znf331        Immune     T Cell             T Cell  T cell (RGCC)   \n",
       "znf442        Stroma       Peri               Peri   Pericyte III   \n",
       "\n",
       "        in_both_panels  \n",
       "aadac            False  \n",
       "abi3bp            True  \n",
       "acan              True  \n",
       "ackr1             True  \n",
       "ackr2             True  \n",
       "...                ...  \n",
       "wnt5a             True  \n",
       "wnt7b             True  \n",
       "xcl1              True  \n",
       "znf331            True  \n",
       "znf442            True  \n",
       "\n",
       "[562 rows x 16 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merfish.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     425\n",
       "False    137\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_genes = merfish.var['in_both_panels'] & merfish.var['highly_variable_intersection']\n",
    "training_genes.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_genes = training_genes[training_genes].index.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For imputation, the scRNAseq object as-is is too large to use as a reference. To ease memory constraints and maximize biologically informative predictions, we need to downsample genes in our reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "scrna = sc.read_h5ad(\"data/normal_skin.scrna.harmony.integrated.reclustered.annotated.mini.h5ad\")\n",
    "## save the correct case over here\n",
    "scrna.var['gene_symbol'] = scrna.var_names.values.tolist().copy()\n",
    "all_genes = scrna.var_names.values.tolist().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "scrna.X = scrna.layers['counts'].copy()\n",
    "sc.pp.normalize_total(scrna)\n",
    "scrna.layers['norm'] = scrna.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "scrna.var_names = scrna.var_names.str.lower().values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 273178 × 5688\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'pct.mito', 'pct.ribo', 'pct.hemo', 'study_id', 'sample_barcode', 'donor_id', 'donor_sex', 'donor_age', 'anatomic_site', 'reported.cell_type', 'anatomic_site.detailed', 'harmony.snn_res.0.2', 'harmony.snn_res.0.4', 'harmony.snn_res.0.5', 'harmony.snn_res.0.6', 'harmony.snn_res.0.8', 'harmony.snn_res.1', 'harmony.snn_res.1.2', 'harmony.snn_res.1.5', 'harmony.snn_res.2', 'harmony.snn_res.2.5', 'seurat_clusters', 'cell_barcode', 'cell_type.broad', 'cell_category', 'cell_type', 'cell_type.broad.res_0.2', 'cell_type.detailed', 'cell_type.reclustered', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'n_counts', 'leiden_0.5', 'leiden_1', 'leiden_1.5', 'leiden_2', 'leiden_2.5'\n",
       "    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'gene_symbol'\n",
       "    uns: 'hvg', 'log1p'\n",
       "    obsm: 'harmony', 'merged_umap', 'pca', 'umap'\n",
       "    layers: 'counts', 'log1p', 'norm'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10433"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PYTORCH_CUDA_ALLOC_CONF={\"expandable_segments\":True}\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_label = \"leiden_2.5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.pyplot import rc_context\n",
    "from sklearn import metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_score_list = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSSM_00\n",
      "Score: 0.112, KL reg: 0.040\n",
      "Score: 0.290, KL reg: 0.003\n",
      "Score: 0.292, KL reg: 0.003\n",
      "Score: 0.292, KL reg: 0.003\n",
      "Score: 0.292, KL reg: 0.003\n",
      "MSSM_01\n",
      "Score: 0.111, KL reg: 0.040\n",
      "Score: 0.297, KL reg: 0.004\n",
      "Score: 0.298, KL reg: 0.003\n",
      "Score: 0.298, KL reg: 0.003\n",
      "Score: 0.298, KL reg: 0.003\n",
      "MSSM_02\n",
      "Score: 0.119, KL reg: 0.040\n",
      "Score: 0.313, KL reg: 0.002\n",
      "Score: 0.314, KL reg: 0.002\n",
      "Score: 0.315, KL reg: 0.002\n",
      "Score: 0.315, KL reg: 0.002\n",
      "MSSM_03\n",
      "Score: 0.141, KL reg: 0.040\n",
      "Score: 0.348, KL reg: 0.003\n",
      "Score: 0.350, KL reg: 0.003\n",
      "Score: 0.350, KL reg: 0.003\n",
      "Score: 0.350, KL reg: 0.003\n",
      "MSSM_04\n",
      "Score: 0.121, KL reg: 0.040\n",
      "Score: 0.313, KL reg: 0.003\n",
      "Score: 0.315, KL reg: 0.003\n",
      "Score: 0.315, KL reg: 0.003\n",
      "Score: 0.315, KL reg: 0.003\n",
      "MSSM_05\n",
      "Score: 0.098, KL reg: 0.040\n",
      "Score: 0.287, KL reg: 0.004\n",
      "Score: 0.288, KL reg: 0.004\n",
      "Score: 0.289, KL reg: 0.004\n",
      "Score: 0.289, KL reg: 0.004\n",
      "MSSM_06\n",
      "Score: 0.110, KL reg: 0.039\n",
      "Score: 0.299, KL reg: 0.003\n",
      "Score: 0.300, KL reg: 0.003\n",
      "Score: 0.300, KL reg: 0.003\n",
      "Score: 0.301, KL reg: 0.003\n",
      "MSSM_07\n",
      "Score: 0.118, KL reg: 0.039\n",
      "Score: 0.312, KL reg: 0.003\n",
      "Score: 0.313, KL reg: 0.003\n",
      "Score: 0.314, KL reg: 0.003\n",
      "Score: 0.314, KL reg: 0.003\n",
      "MSSM_08\n",
      "Score: 0.108, KL reg: 0.040\n",
      "Score: 0.294, KL reg: 0.003\n",
      "Score: 0.295, KL reg: 0.003\n",
      "Score: 0.295, KL reg: 0.003\n",
      "Score: 0.295, KL reg: 0.003\n",
      "MSSM_09\n",
      "Score: 0.116, KL reg: 0.039\n",
      "Score: 0.302, KL reg: 0.003\n",
      "Score: 0.303, KL reg: 0.003\n",
      "Score: 0.304, KL reg: 0.003\n",
      "Score: 0.304, KL reg: 0.003\n",
      "MSSM_10\n",
      "Score: 0.098, KL reg: 0.039\n",
      "Score: 0.274, KL reg: 0.003\n",
      "Score: 0.275, KL reg: 0.003\n",
      "Score: 0.275, KL reg: 0.003\n",
      "Score: 0.275, KL reg: 0.003\n",
      "MSSM_11\n",
      "Score: 0.105, KL reg: 0.040\n",
      "Score: 0.286, KL reg: 0.002\n",
      "Score: 0.288, KL reg: 0.002\n",
      "Score: 0.288, KL reg: 0.002\n",
      "Score: 0.289, KL reg: 0.002\n",
      "MSSM_12\n",
      "Score: 0.106, KL reg: 0.040\n",
      "Score: 0.283, KL reg: 0.003\n",
      "Score: 0.284, KL reg: 0.003\n",
      "Score: 0.285, KL reg: 0.003\n",
      "Score: 0.285, KL reg: 0.003\n",
      "MSSM_13\n",
      "Score: 0.102, KL reg: 0.040\n",
      "Score: 0.287, KL reg: 0.002\n",
      "Score: 0.288, KL reg: 0.002\n",
      "Score: 0.289, KL reg: 0.002\n",
      "Score: 0.289, KL reg: 0.002\n",
      "MSSM_14\n",
      "Score: 0.111, KL reg: 0.039\n",
      "Score: 0.296, KL reg: 0.003\n",
      "Score: 0.297, KL reg: 0.003\n",
      "Score: 0.298, KL reg: 0.003\n",
      "Score: 0.298, KL reg: 0.003\n",
      "MSSM_15\n",
      "Score: 0.124, KL reg: 0.039\n",
      "Score: 0.314, KL reg: 0.003\n",
      "Score: 0.315, KL reg: 0.003\n",
      "Score: 0.316, KL reg: 0.003\n",
      "Score: 0.316, KL reg: 0.003\n",
      "MSSM_16\n",
      "Score: 0.115, KL reg: 0.040\n",
      "Score: 0.309, KL reg: 0.003\n",
      "Score: 0.310, KL reg: 0.003\n",
      "Score: 0.311, KL reg: 0.003\n",
      "Score: 0.311, KL reg: 0.003\n",
      "MSSM_17\n",
      "Score: 0.116, KL reg: 0.040\n",
      "Score: 0.299, KL reg: 0.003\n",
      "Score: 0.301, KL reg: 0.003\n",
      "Score: 0.301, KL reg: 0.003\n",
      "Score: 0.301, KL reg: 0.003\n",
      "MSSM_18\n",
      "Score: 0.108, KL reg: 0.040\n",
      "Score: 0.292, KL reg: 0.003\n",
      "Score: 0.293, KL reg: 0.003\n",
      "Score: 0.293, KL reg: 0.003\n",
      "Score: 0.293, KL reg: 0.003\n",
      "MSSM_19\n",
      "Score: 0.115, KL reg: 0.039\n",
      "Score: 0.294, KL reg: 0.002\n",
      "Score: 0.296, KL reg: 0.002\n",
      "Score: 0.297, KL reg: 0.002\n",
      "Score: 0.297, KL reg: 0.002\n",
      "MSSM_20\n",
      "Score: 0.129, KL reg: 0.040\n",
      "Score: 0.324, KL reg: 0.002\n",
      "Score: 0.326, KL reg: 0.002\n",
      "Score: 0.326, KL reg: 0.002\n",
      "Score: 0.326, KL reg: 0.002\n",
      "MSSM_21\n",
      "Score: 0.132, KL reg: 0.040\n",
      "Score: 0.325, KL reg: 0.003\n",
      "Score: 0.327, KL reg: 0.003\n",
      "Score: 0.327, KL reg: 0.003\n",
      "Score: 0.327, KL reg: 0.003\n",
      "MSSM_22\n",
      "Score: 0.137, KL reg: 0.039\n",
      "Score: 0.330, KL reg: 0.002\n",
      "Score: 0.332, KL reg: 0.002\n",
      "Score: 0.332, KL reg: 0.002\n",
      "Score: 0.332, KL reg: 0.002\n",
      "MSSM_23\n",
      "Score: 0.156, KL reg: 0.040\n",
      "Score: 0.355, KL reg: 0.002\n",
      "Score: 0.357, KL reg: 0.002\n",
      "Score: 0.358, KL reg: 0.002\n",
      "Score: 0.358, KL reg: 0.002\n",
      "MSSM_24\n",
      "Score: 0.113, KL reg: 0.040\n",
      "Score: 0.308, KL reg: 0.003\n",
      "Score: 0.310, KL reg: 0.003\n",
      "Score: 0.310, KL reg: 0.003\n",
      "Score: 0.310, KL reg: 0.003\n"
     ]
    }
   ],
   "source": [
    "for batch_id in merfish_batches:\n",
    "    print(batch_id)\n",
    "    outdir = f\"data/tangram/tangram_outs/\"\n",
    "    os.makedirs(outdir, exist_ok=True)\n",
    "    ad_sp = merfish[merfish.obs['batch'] == batch_id].copy()\n",
    "    sc.pp.filter_genes(ad_sp, min_counts=1)\n",
    "    ad_sp.X = ad_sp.layers['norm'].copy()\n",
    "    gc.collect()\n",
    "\n",
    "    tg.pp_adatas(scrna, ad_sp, genes = training_genes)\n",
    "    gc.collect()\n",
    "    \n",
    "    ad_map = tg.map_cells_to_space(scrna, ad_sp,  \n",
    "                                   density_prior='uniform', \n",
    "                                   device='cuda:0', \n",
    "                                   mode='clusters',\n",
    "                                   num_epochs=500,\n",
    "                                   cluster_label = cluster_label)\n",
    "    ad_map.write_h5ad(f\"{outdir}/{batch_id}.tangram.clusters_mode.mapped_object.5k_mini.h5ad\", compression='gzip')\n",
    "    ad_map.uns['train_genes_df'].to_csv(f'{outdir}/{batch_id}.tangram.clusters_mode.5k_mini.training_stats.csv')\n",
    "    \n",
    "    ad_ge = tg.project_genes(adata_map=ad_map, adata_sc=scrna, cluster_label=cluster_label)\n",
    "    ad_ge.X = sparse.csr_matrix(ad_ge.X, shape=(ad_ge.shape[0], len(all_genes)), dtype=np.float64)\n",
    "    ad_ge.write_h5ad(f\"{outdir}/{batch_id}.tangram.clusters_mode.5k_mini.imputed_object.h5ad\")\n",
    "    \n",
    "    ad_sp.uns['overlap_genes'] = np.sort(ad_sp.uns['overlap_genes'])\n",
    "    ad_ge.uns['overlap_genes'] = np.sort(ad_ge.uns['overlap_genes'])\n",
    "    scrna.uns['overlap_genes'] = np.sort(scrna.uns['overlap_genes'])\n",
    "    \n",
    "    df_all_genes = tg.compare_spatial_geneexp(ad_ge, ad_sp, scrna)\n",
    "    df_all_genes.to_csv(f\"{outdir}/{batch_id}.tangram.clusters_mode.5k_mini.imputation_genestats.csv\")\n",
    "\n",
    "    auc_score = metrics.auc(df_all_genes['score'], df_all_genes['sparsity_sp'])\n",
    "    auc_score_list.append(auc_score)\n",
    "    \n",
    "    ax = sns.regplot(\n",
    "        data=df_all_genes, \n",
    "        x=\"score\", y=\"sparsity_sp\", \n",
    "        logistic=True, color=\".3\", \n",
    "        line_kws=dict(color=\"r\"),\n",
    "        label = 'AUC = %0.2f' % auc_score\n",
    "    )\n",
    "    ax.set_title(batch_id + ' (AUC = %0.2f' % auc_score + \")\")\n",
    "    plt.savefig(f\"figures/tangram/qc/{batch_id}.tangram.clusters_mode.5k_mini.imputation.auc_curve.png\")\n",
    "    plt.close()\n",
    "    \n",
    "    torch.cuda.empty_cache()\n",
    "    gc.collect()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71329"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ge.var_names = ad_ge.var['gene_symbol'].values.tolist().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>ribo</th>\n",
       "      <th>hb</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>highly_variable_nbatches</th>\n",
       "      <th>highly_variable_intersection</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>sparsity</th>\n",
       "      <th>is_training</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>27896</td>\n",
       "      <td>0.630981</td>\n",
       "      <td>89.788343</td>\n",
       "      <td>172370.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.135896</td>\n",
       "      <td>0.891595</td>\n",
       "      <td>-0.062615</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>27896</td>\n",
       "      <td>0.897883</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>25439</td>\n",
       "      <td>0.434189</td>\n",
       "      <td>90.687757</td>\n",
       "      <td>118611.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.313068</td>\n",
       "      <td>2.187632</td>\n",
       "      <td>1.869681</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>A2M</td>\n",
       "      <td>25439</td>\n",
       "      <td>0.906878</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1337</td>\n",
       "      <td>0.087961</td>\n",
       "      <td>99.510576</td>\n",
       "      <td>24029.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.007811</td>\n",
       "      <td>1.071408</td>\n",
       "      <td>0.299354</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>1337</td>\n",
       "      <td>0.995106</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7791</td>\n",
       "      <td>0.089645</td>\n",
       "      <td>97.148013</td>\n",
       "      <td>24489.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.029538</td>\n",
       "      <td>1.643422</td>\n",
       "      <td>1.450835</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>7791</td>\n",
       "      <td>0.971480</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>13819</td>\n",
       "      <td>0.078143</td>\n",
       "      <td>94.941394</td>\n",
       "      <td>21347.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.078814</td>\n",
       "      <td>1.068986</td>\n",
       "      <td>0.294479</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>13819</td>\n",
       "      <td>0.949414</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>96</td>\n",
       "      <td>0.000780</td>\n",
       "      <td>99.964858</td>\n",
       "      <td>213.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>1.126660</td>\n",
       "      <td>0.410579</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>ZSCAN4</td>\n",
       "      <td>96</td>\n",
       "      <td>0.999649</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7241</td>\n",
       "      <td>0.041592</td>\n",
       "      <td>97.349347</td>\n",
       "      <td>11362.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.029834</td>\n",
       "      <td>0.972179</td>\n",
       "      <td>0.099604</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>7241</td>\n",
       "      <td>0.973493</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM5</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1267</td>\n",
       "      <td>0.020917</td>\n",
       "      <td>99.536200</td>\n",
       "      <td>5714.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.007878</td>\n",
       "      <td>1.470241</td>\n",
       "      <td>1.102215</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>ZSWIM5</td>\n",
       "      <td>1267</td>\n",
       "      <td>0.995362</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM6</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>31720</td>\n",
       "      <td>0.233932</td>\n",
       "      <td>88.388523</td>\n",
       "      <td>63905.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.154845</td>\n",
       "      <td>1.033158</td>\n",
       "      <td>0.222356</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>31720</td>\n",
       "      <td>0.883885</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>22649</td>\n",
       "      <td>0.394728</td>\n",
       "      <td>91.709069</td>\n",
       "      <td>107831.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.095302</td>\n",
       "      <td>0.975789</td>\n",
       "      <td>0.106870</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>22649</td>\n",
       "      <td>0.917091</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5688 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            mt   ribo     hb  n_cells_by_counts  mean_counts  \\\n",
       "A1BG     False  False  False              27896     0.630981   \n",
       "A2M      False  False  False              25439     0.434189   \n",
       "A2M-AS1  False  False  False               1337     0.087961   \n",
       "A2ML1    False  False  False               7791     0.089645   \n",
       "A4GALT   False  False  False              13819     0.078143   \n",
       "...        ...    ...    ...                ...          ...   \n",
       "ZSCAN4   False  False  False                 96     0.000780   \n",
       "ZSWIM4   False  False  False               7241     0.041592   \n",
       "ZSWIM5   False  False  False               1267     0.020917   \n",
       "ZSWIM6   False  False  False              31720     0.233932   \n",
       "ZZEF1    False  False  False              22649     0.394728   \n",
       "\n",
       "         pct_dropout_by_counts  total_counts  highly_variable     means  \\\n",
       "A1BG                 89.788343      172370.0            False  0.135896   \n",
       "A2M                  90.687757      118611.0             True  0.313068   \n",
       "A2M-AS1              99.510576       24029.0             True  0.007811   \n",
       "A2ML1                97.148013       24489.0             True  0.029538   \n",
       "A4GALT               94.941394       21347.0             True  0.078814   \n",
       "...                        ...           ...              ...       ...   \n",
       "ZSCAN4               99.964858         213.0             True  0.000426   \n",
       "ZSWIM4               97.349347       11362.0             True  0.029834   \n",
       "ZSWIM5               99.536200        5714.0             True  0.007878   \n",
       "ZSWIM6               88.388523       63905.0             True  0.154845   \n",
       "ZZEF1                91.709069      107831.0             True  0.095302   \n",
       "\n",
       "         dispersions  dispersions_norm  highly_variable_nbatches  \\\n",
       "A1BG        0.891595         -0.062615                        14   \n",
       "A2M         2.187632          1.869681                        14   \n",
       "A2M-AS1     1.071408          0.299354                        14   \n",
       "A2ML1       1.643422          1.450835                        14   \n",
       "A4GALT      1.068986          0.294479                        14   \n",
       "...              ...               ...                       ...   \n",
       "ZSCAN4      1.126660          0.410579                        11   \n",
       "ZSWIM4      0.972179          0.099604                        14   \n",
       "ZSWIM5      1.470241          1.102215                        14   \n",
       "ZSWIM6      1.033158          0.222356                        14   \n",
       "ZZEF1       0.975789          0.106870                        14   \n",
       "\n",
       "         highly_variable_intersection gene_symbol  n_cells  sparsity  \\\n",
       "A1BG                             True        A1BG    27896  0.897883   \n",
       "A2M                              True         A2M    25439  0.906878   \n",
       "A2M-AS1                          True     A2M-AS1     1337  0.995106   \n",
       "A2ML1                            True       A2ML1     7791  0.971480   \n",
       "A4GALT                           True      A4GALT    13819  0.949414   \n",
       "...                               ...         ...      ...       ...   \n",
       "ZSCAN4                          False      ZSCAN4       96  0.999649   \n",
       "ZSWIM4                           True      ZSWIM4     7241  0.973493   \n",
       "ZSWIM5                           True      ZSWIM5     1267  0.995362   \n",
       "ZSWIM6                           True      ZSWIM6    31720  0.883885   \n",
       "ZZEF1                            True       ZZEF1    22649  0.917091   \n",
       "\n",
       "         is_training  \n",
       "A1BG           False  \n",
       "A2M            False  \n",
       "A2M-AS1        False  \n",
       "A2ML1          False  \n",
       "A4GALT         False  \n",
       "...              ...  \n",
       "ZSCAN4         False  \n",
       "ZSWIM4         False  \n",
       "ZSWIM5         False  \n",
       "ZSWIM6         False  \n",
       "ZZEF1          False  \n",
       "\n",
       "[5688 rows x 17 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad_ge.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gc\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/tangram/tangram_outs/MSSM_19.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_11.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_02.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_04.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_21.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_24.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_23.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_22.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_16.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_08.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_12.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_17.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_05.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_13.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_09.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_00.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_07.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_20.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_18.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_14.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_15.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_01.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_06.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_10.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n",
      "data/tangram/tangram_outs/MSSM_03.tangram.clusters_mode.5k_mini.imputed_object.h5ad\n"
     ]
    }
   ],
   "source": [
    "flist = glob.glob(\"data/tangram/tangram_outs/*.clusters_mode.5k_mini.imputed_object.h5ad\")\n",
    "adata_list = []\n",
    "for f in flist:\n",
    "    print(f)\n",
    "    sample_adata = ad.read_h5ad(f, backed='r')\n",
    "    adata_list.append(sample_adata)\n",
    "    del sample_adata\n",
    "    gc.collect()\n",
    "gc.collect()\n",
    "\n",
    "from anndata.experimental.multi_files import AnnCollection\n",
    "dataset = AnnCollection(adata_list, join_vars='inner', join_obs='outer', join_obsm='inner', label='dataset')\n",
    "adata = dataset[dataset.obs_names].to_adata()\n",
    "del dataset\n",
    "gc.collect()\n",
    "\n",
    "adata.obsm['spatial'] =  adata.obs[['center_x', 'center_y']].values.copy()\n",
    "adata.layers['tangram'] = adata.X.copy()\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "adata.layers['scaled'] = adata.X.copy()\n",
    "gc.collect()\n",
    "adata.write_h5ad(\"data/tangram/merfish.tangram_imputed.5k_mini.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names = scrna.var['gene_symbol'].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(\"data/tangram/merfish.tangram_imputed.5k_mini.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"data/tangram/merfish.tangram_imputed.5k_mini.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from anndata import AnnData\n",
    "\n",
    "def minmax_normalize_by_gene_per_sample(adata: AnnData, sample_key: str = 'sample_barcode', layer_name: str = 'minmax') -> None:\n",
    "    \"\"\"\n",
    "    Min-max normalize gene expression per gene (column) within each sample group,\n",
    "    and store the result in adata.layers[layer_name].\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : AnnData\n",
    "        AnnData object with .X as dense or sparse array.\n",
    "    sample_key : str\n",
    "        Column in adata.obs identifying the sample grouping (e.g., 'sample_barcode').\n",
    "    layer_name : str\n",
    "        Name of the .layers entry to store the result (default: 'minmax').\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "        Modifies adata in-place by writing to adata.layers[layer_name].\n",
    "    \"\"\"\n",
    "    from scipy.sparse import issparse\n",
    "\n",
    "    if issparse(adata.X):\n",
    "        X = adata.X.toarray()\n",
    "    else:\n",
    "        X = adata.X.copy()\n",
    "\n",
    "    norm_X = np.zeros_like(X)\n",
    "    sample_barcodes = adata.obs[sample_key].unique()\n",
    "\n",
    "    for barcode in sample_barcodes:\n",
    "        mask = adata.obs[sample_key].values == barcode\n",
    "        sub_X = X[mask]  # shape: [cells_in_sample, genes]\n",
    "\n",
    "        min_vals = sub_X.min(axis=0, keepdims=True)  # shape: [1, genes]\n",
    "        max_vals = sub_X.max(axis=0, keepdims=True)  # shape: [1, genes]\n",
    "\n",
    "        denom = np.where((max_vals - min_vals) == 0, 1, max_vals - min_vals)\n",
    "        norm_X[mask] = (sub_X - min_vals) / denom\n",
    "\n",
    "    adata.layers[layer_name] = norm_X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "minmax_normalize_by_gene_per_sample(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "84193"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(\"data/tangram/merfish.tangram_imputed.5k_mini.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_genes = [\"KRT1\", \"KRT5\", \"KRT14\", \"KRT10\", \"KRT15\", \"KRT17\", \n",
    "              \"DSG1\", \"KRTDAP\", \"DMKN\", \"SBSN\", \"FABP5\", \"COL1A2\",\n",
    "              \"SCGB1B2P\", \"KRT23\", \"MGST1\", \"KRT19\", \"TNFSFR1A\", \"TNF\", \n",
    "              'CXCL12', \"CXCR4\", 'PECAM1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.colors as clr\n",
    "gray2blue = sns.blend_palette([sns.xkcd_rgb[\"light grey\"], sns.xkcd_rgb[\"electric blue\"]], as_cmap=True)\n",
    "gray2red = sns.blend_palette([sns.xkcd_rgb[\"light grey\"], sns.xkcd_rgb[\"bright red\"]], as_cmap=True)\n",
    "gray2teal = sns.blend_palette([sns.xkcd_rgb[\"light grey\"], sns.xkcd_rgb[\"turquoise\"]], as_cmap=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes = [\"KRT1\", \"KRT5\", \"KRT14\", \"KRT10\", \"KRT15\", \"KRT17\", \n",
    "                \"DSG1\", \"KRTDAP\", \"DMKN\", \"SBSN\", \"FABP5\", \"COL1A2\",\n",
    "                \"SCGB1B2P\", \"KRT23\", \"MGST1\", \"KRT19\", \"COL1A1\", \n",
    "                \"KRT79\", \"ADIPOQ\", \"MRC1\", \"CCL19\", \"APOE\", \n",
    "                \"CD4\", \"PDGFRA\", \"FLG\", \"ACKR1\", \"CLDN5\", \"VWF\", \"VIM\",\n",
    "               \"PLVAP\", \"PODXL\", \"PROX1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_genes = [\"CD74\", \"CD44\", \"BSG\", \"TNFRSF1A\", \"TNFRSF1B\", \"CXCR4\", \"ITGB1\", \"ITGA3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "receptor_genes = [\"COMP\", \"CXCL12\", \"MIF\", \"TNF\", \"PPIA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "def plot_spatial_featureplot_by_sample(\n",
    "    adata,\n",
    "    gene_key,\n",
    "    sample_key=\"sample_barcode\",\n",
    "    layer=None,\n",
    "    cmap=\"magma\",\n",
    "    figsize=(5, 5),    \n",
    "    output_directory='.',\n",
    "    file_prefix='',\n",
    "    dpi=300,\n",
    "    show=False\n",
    "):\n",
    "    samples = adata.obs[sample_key].cat.categories.unique().sort_values()\n",
    "    n_samples = len(samples)\n",
    "\n",
    "    ncols = min(n_samples, 6)\n",
    "    nrows = (n_samples + ncols - 1) // ncols\n",
    "\n",
    "    if layer != None:\n",
    "        adata.X = adata.layers[layer]\n",
    "\n",
    "    fig, axes = plt.subplots(nrows, ncols, figsize=(figsize[0]*ncols, figsize[1]*nrows))\n",
    "    axes = np.array(axes).reshape(-1)\n",
    "\n",
    "    for i, sample in enumerate(samples):\n",
    "        ax = axes[i]\n",
    "        adata_sub = adata[adata.obs[sample_key] == sample]\n",
    "        ax.set_aspect('equal', adjustable='datalim')\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "        sc.pl.embedding(adata_sub, \n",
    "                        basis='spatial',\n",
    "                        title=f\"{sample}\", \n",
    "                        color=gene_key,\n",
    "                        cmap=cmap,\n",
    "                        ax=ax, \n",
    "                        legend_loc='on data', \n",
    "                        show=False)\n",
    "        \n",
    "        ax.axis(\"off\")\n",
    "\n",
    "    for j in range(i + 1, len(axes)):\n",
    "        axes[j].set_visible(False)\n",
    "\n",
    "    # Add a global panel title\n",
    "    fig.suptitle(f\"{gene_key}\", fontsize=14, y=0.99)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{output_directory}/{file_prefix}.{gene_key}.spatial_plot.png\", \n",
    "                bbox_inches='tight', dpi=dpi)\n",
    "    if show:\n",
    "        plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|█████████████████████████████████████████████| 32/32 [28:14<00:00, 52.94s/Gene]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "with tqdm(total=len(marker_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in marker_genes:\n",
    "        if gene in adata.var_names:\n",
    "            plot_spatial_featureplot_by_sample(\n",
    "               adata,\n",
    "               output_directory=\"figures/tangram/feature_spatial/\",\n",
    "               file_prefix=f'tangram_imputed.minmax',\n",
    "               layer='minmax',\n",
    "               gene_key=gene,\n",
    "               cmap=gray2red,\n",
    "               sample_key='sample_barcode')\n",
    "            pbar.update(1)\n",
    "            gc.collect()\n",
    "    gc.collect()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|█████████████████████████████████████████████| 32/32 [28:35<00:00, 53.60s/Gene]\n"
     ]
    }
   ],
   "source": [
    "with tqdm(total=len(marker_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in marker_genes:\n",
    "        if gene in adata.var_names:\n",
    "            plot_spatial_featureplot_by_sample(\n",
    "               adata,\n",
    "               output_directory=\"figures/tangram/feature_spatial/\",\n",
    "               file_prefix=f'tangram_imputed.scaled',\n",
    "               layer='scaled',\n",
    "               gene_key=gene, \n",
    "               cmap=gray2red,\n",
    "               sample_key='sample_barcode')\n",
    "            pbar.update(1)\n",
    "            gc.collect()\n",
    "    gc.collect()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|███████████████████████████████████████████████| 8/8 [06:44<00:00, 50.52s/Gene]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "with tqdm(total=len(ligand_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in ligand_genes:\n",
    "        plot_spatial_featureplot_by_sample(\n",
    "           adata,\n",
    "           output_directory=\"figures/tangram/feature_spatial/\",\n",
    "           file_prefix=f'tangram_imputed.minmax',\n",
    "           layer='minmax',\n",
    "           gene_key=gene,\n",
    "           cmap=gray2red,\n",
    "           sample_key='sample_barcode')\n",
    "        pbar.update(1)\n",
    "        gc.collect()\n",
    "    gc.collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|███████████████████████████████████████████████| 8/8 [06:50<00:00, 51.33s/Gene]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "with tqdm(total=len(ligand_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in ligand_genes:\n",
    "        plot_spatial_featureplot_by_sample(\n",
    "           adata,\n",
    "           output_directory=\"figures/tangram/feature_spatial/\",\n",
    "           file_prefix=f'tangram_imputed.scaled',\n",
    "           layer='scaled',\n",
    "           gene_key=gene,\n",
    "           cmap=gray2red,\n",
    "           sample_key='sample_barcode')\n",
    "        pbar.update(1)\n",
    "        gc.collect()\n",
    "    gc.collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|███████████████████████████████████████████████| 5/5 [04:15<00:00, 51.16s/Gene]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "with tqdm(total=len(receptor_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in receptor_genes:\n",
    "        plot_spatial_featureplot_by_sample(\n",
    "           adata,\n",
    "           output_directory=\"figures/tangram/feature_spatial/\",\n",
    "           file_prefix=f'tangram_imputed.minmax',\n",
    "           layer='minmax',\n",
    "           gene_key=gene,\n",
    "           cmap=gray2blue,\n",
    "           sample_key='sample_barcode')\n",
    "        pbar.update(1)\n",
    "        gc.collect()\n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Genes: 100%|███████████████████████████████████████████████| 5/5 [04:14<00:00, 50.90s/Gene]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "with tqdm(total=len(receptor_genes), desc=\"Plotting Genes\", unit=\"Gene\", ncols=100, leave=True) as pbar:\n",
    "    for gene in receptor_genes:\n",
    "        plot_spatial_featureplot_by_sample(\n",
    "           adata,\n",
    "           output_directory=\"figures/tangram/feature_spatial/\",\n",
    "           file_prefix=f'tangram_imputed.scaled',\n",
    "           layer='scaled',\n",
    "           gene_key=gene,\n",
    "           cmap=gray2blue,\n",
    "           sample_key='sample_barcode')\n",
    "        pbar.update(1)\n",
    "        gc.collect()\n",
    "    gc.collect()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
